spiner/piecewise_grid_1d.hpp

```diff
@@ -0,0 +1,220 @@
+#ifndef SPINER_HIERARCHICAL_GRID_1D_
+#define SPINER_HIERARCHICAL_GRID_1D_
+//======================================================================
+// © (or copyright) 2019-2023. Triad National Security, LLC. All rights
+// reserved.  This program was produced under U.S. Government contract
+// 89233218CNA000001 for Los Alamos National Laboratory (LANL), which is
+// operated by Triad National Security, LLC for the U.S.  Department of
+// Energy/National Nuclear Security Administration. All rights in the
+// program are reserved by Triad National Security, LLC, and the
+// U.S. Department of Energy/National Nuclear Security
+// Administration. The Government is granted for itself and others acting
+// on its behalf a nonexclusive, paid-up, irrevocable worldwide license
+// in this material to reproduce, prepare derivative works, distribute
+// copies to the public, perform publicly and display publicly, and to
+// permit others to do so.
+//======================================================================
+
+#include <array>
+#include <initializer_list>
+#include <limits>
+#include <utility>
+#include <vector>
+
+#include <ports-of-call/portability.hpp>
+#include <ports-of-call/portable_errors.hpp>
+
+#include "regular_grid_1d.hpp"
+
+namespace Spiner {
+
+template <typename T = Real, int NGRIDSMAX = 5,
+          typename =
+              typename std::enable_if<std::is_arithmetic<T>::value, bool>::type>
+class PiecewiseGrid1D {
+ public:
+  using ValueType = T;
+  static constexpr int BAD_VALUE = -1;
+
+  // __host__ __device__ default constructors cause warning.
+  // This is functionally equivalent because grids_ will
+  // be initialized to default values
+  PORTABLE_INLINE_FUNCTION PiecewiseGrid1D() {}
+  PiecewiseGrid1D(const std::vector<RegularGrid1D<T>> grids) {
+    NGRIDS_ = grids.size();
+    PORTABLE_ALWAYS_REQUIRE(
+        NGRIDS_ <= NGRIDSMAX,
+        "Total number of grids must be within maximum allowed");
+    int point_tot = 0;
+    for (int i = 0; i < NGRIDS_; ++i) {
+      grids_[i] = grids[i];
+      pointTotals_[i] = point_tot;
+      point_tot += grids[i].nPoints();
+      if ((i > 0) &&
+          ratio_(2 * std::abs(grids_[i].min() - grids_[i - 1].max()),
+                 std::abs(grids_[i].min() + grids_[i - 1].max()) >= EPS_())) {
```

Comment:
```suggestion
                 std::abs(grids_[i].min() + grids_[i - 1].max())) >= EPS_()) {
```
I think this fixes the parentheses issue

Reply:
I'm a bit confused why the template deduction doesn't fail, but I literally copy-pasted this code into compiler explorer and the template deduction fails there.

Reply:
Yes I think you're correct. Need to move the `)` to after the denomintaor's `max` to close the ratio function.

Reply:
I think it is doing some implicit conversions to make the code legal.

Reply:
Fixed and swapped to a more readable version of the conditional. Thanks for the fix @jhp-lanl